### ISSUE TITLE
SSL_is_init_finished() returns nonzero before SSL_read_early_data() returns SSL_READ_EARLY_DATA_FINISH 

### ISSUE THREAD

**Issue by tatsuhiro-t** [openssl/openssl#3041]
I saw that SSL_is_init_finished() returns nonzero before SSL_read_early_data() returns SSL_READ_EARLY_DATA_FINISH on my experimental TLSv1.3 deployment.
Is this expected behaviour?

I use SSL_is_init_finished() to select write function between SSL_write_early_data and SSL_write.  It seems that I should also consider that SSL_read_eary_data has returned SSL_READ_EARLY_DATA_FINISH already.

Here is some trace when this happened.
First, the server read 42 bytes by SSL_read_early_data.
Then the server read 5 bytes, and 32 bytes of early data using SSL_read_early_data. At that point, it still didn't return SSL_READ_EARLY_DATA_FINISH.
I didn't log each write function.  So I don't know the exact timing and which function was used.
Eventually, I got: "error:1420C042:SSL routines:ssl_write_internal:called a function you should not call" by calling SSL_write().  In the server code, use of SSL_write is used only when SSL_is_init_finished() returns nonzero.

**Comment by mattcaswell** [comment]
No - this is not expected behaviour. Are you able to construct a simple reproducing test case?

**Comment by tatsuhiro-t** [comment]
Will try.  My guess is that several early data fragments in one packet may cause this issue, but I could be wrong.

**Comment by tatsuhiro-t** [comment]
I'm struggling to make a test case.
I found the log trace that ticket callback set by SSL_CTX_set_tlsext_ticket_key_cb was called before SSL_read_early_data returned SSL_READ_EARLY_DATA_FINISH.  Does it help debugging this issue?

**Comment by mattcaswell** [comment]
Not really. Are you able to replicate this issue in your own application, or was this a one-off event?

**Comment by tatsuhiro-t** [comment]
No, I couldn't reproduce this issue with my application.  I tried to write a test client using OpenSSL master, but it didn't cause any issue.

It is not one off really.  I observed this exact same issue 6 times.  From the HTTP request, they are the same, so I think they came from the same endpoint.  In that period, IETF hackathon was going on, non-openssl implementation could cause this issue, but I'm really not sure.

**Comment by mattcaswell** [comment]
Are you running the older draft-18 implementation or the latest master (i.e. draft-19)?

**Comment by tatsuhiro-t** [comment]
Latest master, draft-19.  I think b6ef12c4baa3a2c1ff0e3ac71270588dfcfe8cbd was used.

**Comment by mattcaswell** [comment]
I've spent quite a bit of time going through the code trying to think of a scenario where this could occur, and I can't think of one. To make progress I think we're going to need to get more logging info. I am particularly interested in the state of various things both before and after calls to SSL_read_early_data()/SSL_write_early_data()/SSL_is_init_finished() etc. 
The info I'd like to know at those points is:
- the value of the "early_data_state" member in the SSL struct
- the value returned from a call to SSL_get_state()
- the return value from SSL_read_early_data()/SSL_write_early_data()/SSL_is_init_finished() etc. after calls to those functions as well as the amount of data read/written

"early_data_state" is not exposed by any form of "getter" so the only way you can access it is by including the ssl/ssl_locl.h header file in your application (obviously not recommended for normal code - but the for the purposes of debugging - ok). Alternately I could pull together a temporary patch to provide an accessor.

Would you be able to add such logging?

**Comment by tatsuhiro-t** [comment]
Yes.  I'll do this tonight, and deploy the server.

**Comment by tatsuhiro-t** [comment]
I got debug log, but it shows the issue in the other way around; that is I got "SSL_write_early_data:called a function you should not call" when SSL_is_init_finished returned 0.
I'll try to reproduce the original issue.  Anyway, here is the trace:

```
tls: read 553 bytes
DEBUG: func=SSL_read_early_data(before), ssl->early_data_state=0, SSL_get_state()=0
decrypt session ticket key: c5e2137a43a3ee0ffa81f2a7ee8f5dd2
DEBUG: func=SSL_read_early_data(after), ssl->early_data_state=9, SSL_get_state()=48
DEBUG: SSL_read_eary_data()=1, nread=42
DEBUG: func=SSL_read_early_data(before), ssl->early_data_state=9, SSL_get_state()=48
DEBUG: func=SSL_read_early_data(after), ssl->early_data_state=9, SSL_get_state()=48
DEBUG: SSL_read_eary_data()=0, nread=0
DEBUG: func=SSL_read_early_data(before), ssl->early_data_state=9, SSL_get_state()=48
DEBUG: func=SSL_read_early_data(after), ssl->early_data_state=9, SSL_get_state()=48
DEBUG: SSL_read_eary_data()=0, nread=0
DEBUG: func=SSL_read_early_data(before), ssl->early_data_state=9, SSL_get_state()=48
DEBUG: func=SSL_read_early_data(after), ssl->early_data_state=9, SSL_get_state()=48
DEBUG: SSL_read_eary_data()=0, nread=0
DEBUG: func=SSL_read_early_data(before), ssl->early_data_state=9, SSL_get_state()=48
tls: read 5 bytes
tls: read 21 bytes
DEBUG: func=SSL_is_init_finished, ssl->early_data_state=10, SSL_get_state()=48
DEBUG: SSL_is_init_finished()=0
DEBUG: func=SSL_read_early_data(after), ssl->early_data_state=11, SSL_get_state()=51
DEBUG: SSL_read_eary_data()=2, nread=0
DEBUG: func=SSL_is_init_finished, ssl->early_data_state=11, SSL_get_state()=51
DEBUG: SSL_is_init_finished()=0
DEBUG: func=SSL_write_early_data(before), ssl->early_data_state=11, SSL_get_state()=51
DEBUG: func=SSL_write_early_data(after), ssl->early_data_state=11, SSL_get_state()=51
DEBUG: SSL_write_eary_data()=0, nwrite=0
SSL_write: error:1420E042:SSL routines:SSL_write_early_data:called a function you should not call
```

`tls: read NN bytes` means that server read NN bytes from TCP socket.

**Comment by tatsuhiro-t** [comment]
Here is the trace which matches the original report:

```
tls: read 489 bytes
DEBUG: func=SSL_read_early_data(before), ssl->early_data_state=0, SSL_get_state()=0
decrypt session ticket key: 7e9a06ce1963c921775ac6eca811d2a1
DEBUG: func=SSL_read_early_data(after), ssl->early_data_state=9, SSL_get_state()=48
DEBUG: SSL_read_eary_data()=0, nread=0
DEBUG: func=SSL_read_early_data(before), ssl->early_data_state=9, SSL_get_state()=48
DEBUG: func=SSL_read_early_data(after), ssl->early_data_state=9, SSL_get_state()=48
DEBUG: SSL_read_eary_data()=0, nread=0
tls: read 64 bytes
DEBUG: func=SSL_read_early_data(before), ssl->early_data_state=9, SSL_get_state()=48
DEBUG: func=SSL_read_early_data(after), ssl->early_data_state=9, SSL_get_state()=48
DEBUG: SSL_read_eary_data()=1, nread=42
DEBUG: func=SSL_read_early_data(before), ssl->early_data_state=9, SSL_get_state()=48
DEBUG: func=SSL_read_early_data(after), ssl->early_data_state=9, SSL_get_state()=48
DEBUG: SSL_read_eary_data()=0, nread=0
DEBUG: func=SSL_read_early_data(before), ssl->early_data_state=9, SSL_get_state()=48
DEBUG: func=SSL_read_early_data(after), ssl->early_data_state=9, SSL_get_state()=48
DEBUG: SSL_read_eary_data()=0, nread=0
DEBUG: func=SSL_read_early_data(before), ssl->early_data_state=9, SSL_get_state()=48
DEBUG: func=SSL_read_early_data(after), ssl->early_data_state=9, SSL_get_state()=48
DEBUG: SSL_read_eary_data()=0, nread=0
DEBUG: func=SSL_read_early_data(before), ssl->early_data_state=9, SSL_get_state()=48
tls: read 5 bytes
tls: read 21 bytes
DEBUG: func=SSL_is_init_finished, ssl->early_data_state=10, SSL_get_state()=48
DEBUG: SSL_is_init_finished()=0
tls: read 5 bytes
tls: read 69 bytes
encrypt session ticket key: 7e9a06ce1963c921775ac6eca811d2a1
tls: read 5 bytes
tls: read 49 bytes
DEBUG: func=SSL_read_early_data(after), ssl->early_data_state=9, SSL_get_state()=1
DEBUG: SSL_read_eary_data()=1, nread=32
DEBUG: func=SSL_read_early_data(before), ssl->early_data_state=9, SSL_get_state()=1
tls: read 5 bytes
tls: read 20 bytes
DEBUG: func=SSL_read_early_data(after), ssl->early_data_state=9, SSL_get_state()=1
DEBUG: SSL_read_eary_data()=1, nread=3
DEBUG: func=SSL_read_early_data(before), ssl->early_data_state=9, SSL_get_state()=1
DEBUG: func=SSL_read_early_data(after), ssl->early_data_state=9, SSL_get_state()=1
DEBUG: SSL_read_eary_data()=0, nread=0
DEBUG: func=SSL_is_init_finished, ssl->early_data_state=9, SSL_get_state()=1
DEBUG: SSL_is_init_finished()=1
error:1420C042:SSL routines:ssl_write_internal:called a function you should not call
```

The interesting thing in this trace is that "DEBUG: SSL_read_eary_data()=1, nread=32" and "DEBUG: SSL_read_eary_data()=1, nread=3" seem to read the non-early data.  I have to ask the author of the client, but I only sent 42 bytes data in early data which has already been read before.

**Comment by mattcaswell** [comment]
Are you using read_ahead? (e.g. via SSL_CTX_set_read_ahead() or SSL_set_read_ahead())

**Comment by tatsuhiro-t** [comment]
No.  I don't use either of them.

**Comment by tatsuhiro-t** [comment]
BTW, some of "tls: read NN bytes" is logged in custom bio read function.

**Comment by mattcaswell** [comment]
What about SSL_MODE_AUTO_RETRY?

**Comment by mattcaswell** [comment]
i.e. I mean are you using SSL_MODE_AUTO_RETRY?

**Comment by tatsuhiro-t** [comment]
Yes.  I use SSL_MODE_AUTO_RETRY.

**Comment by mattcaswell** [comment]
Aha! I think SSL_read_early_data() is attempting to read data but hitting the EndOfEarlyData message. Before it can return control to the application with SSL_READ_EARLY_DATA_FINISH it is auto retrying - but this time reading normal data!

**Comment by mattcaswell** [comment]
Please can you try out the patch in #3077

**Comment by tatsuhiro-t** [comment]
I confirmed that the patch #3077 fixes the original issue.  Thank you!

However, I still encounter the bug described in https://github.com/openssl/openssl/issues/3041#issuecomment-290084587, where I got "error:1420E042:SSL routines:SSL_write_early_data:called a function you should not call" even if SSL_is_init_finished() returned 0.

**Comment by mattcaswell** [comment]
Ok. The first fix is now in master. I'll take a look at the other problem.

**Comment by mattcaswell** [comment]
This issue seems to occur when there is a pause between the arrival of EoED and CF. In that case SSL_read_early_data() will return SSL_READ_EARLY_DATA_FINISH but we still haven't finished the handshake yet. Fix in #3089.

**Comment by mattcaswell** [comment]
The second fix is now in master, so hopefully this should be resolved now. Let me know if not.

**Comment by tatsuhiro-t** [comment]
Thank you.  With the latest master, I got a bit different situation.
Now SSL_is_init_finished(ssl)=0, and SSL_wirte_early_data() is called.  But it returns 0, and the error code is SSL_ERROR_WANT_READ.

**Comment by tatsuhiro-t** [comment]
In other words, I expect that SSL_write_early_data() never return SSL_ERROR_WANT_READ.  This is because, according to the manual, if SSL_is_init_finished(ssl) returns 0, it means I can still write early unauthenticated data without waiting for another read.

**Comment by mattcaswell** [comment]
Grrrrr.....ok....looking at it.

**Comment by mattcaswell** [comment]
Third time lucky. New fix in #3091. The earlier test I introduced for your second problem (delayed arrival of CF), didn't quite push back the arrival late enough to expose this new issue.

**Comment by tatsuhiro-t** [comment]
Unfortunately, with the patch proposed in #3091, I got the following trace:

```
tls: read 553 bytes
DEBUG: func=SSL_read_early_data(before), ssl->early_data_state=0, SSL_get_state()=0
decrypt session ticket key: 09a3c294712c006f4fa5a8b371fd6ec0
DEBUG: func=SSL_read_early_data(after), ssl->early_data_state=9, SSL_get_state()=48
DEBUG: SSL_read_eary_data()=1, nread=42
DEBUG: func=SSL_read_early_data(before), ssl->early_data_state=9, SSL_get_state()=48
DEBUG: func=SSL_read_early_data(after), ssl->early_data_state=9, SSL_get_state()=48
DEBUG: SSL_read_eary_data()=0, nread=0
DEBUG: func=SSL_read_early_data(before), ssl->early_data_state=9, SSL_get_state()=48
DEBUG: func=SSL_read_early_data(after), ssl->early_data_state=9, SSL_get_state()=48
DEBUG: SSL_read_eary_data()=0, nread=0
DEBUG: func=SSL_read_early_data(before), ssl->early_data_state=9, SSL_get_state()=48
tls: read 5 bytes
tls: read 21 bytes
DEBUG: func=SSL_is_init_finished, ssl->early_data_state=10, SSL_get_state()=48
DEBUG: SSL_is_init_finished()=0
DEBUG: func=SSL_read_early_data(after), ssl->early_data_state=11, SSL_get_state()=51
DEBUG: SSL_read_eary_data()=2, nread=0
DEBUG: func=SSL_is_init_finished, ssl->early_data_state=11, SSL_get_state()=51
DEBUG: SSL_is_init_finished()=0
DEBUG: func=SSL_write_early_data(before), ssl->early_data_state=11, SSL_get_state()=51
DEBUG: func=SSL_write_early_data(after), ssl->early_data_state=9, SSL_get_state()=51
DEBUG: SSL_write_eary_data()=1, nwrite=431
tls: read 5 bytes
tls: read 69 bytes
encrypt session ticket key: 09a3c294712c006f4fa5a8b371fd6ec0
tls: read 5 bytes
tls: read 49 bytes
tls: read 5 bytes
tls: read 20 bytes
DEBUG: func=SSL_is_init_finished, ssl->early_data_state=9, SSL_get_state()=1
DEBUG: SSL_is_init_finished()=1
SSL_write: error:1420C042:SSL routines:ssl_write_internal:called a function you should not call
```

It seems there is a progress.  But unlocked next hurdle.  Now SSL_write shouted error even if SSL_is_init_finished() returned nonzero.

**Comment by mattcaswell** [comment]
I will investigate further. Note though that I am not going to be around much for the rest of today and until the end of the weekend. It might be Monday before I can get to it.

**Comment by tatsuhiro-t** [comment]
There is no rush.
BTW, I'm going to remove SSL_MODE_AUTO_RETRY from my server since it uses non-blocking I/O.
I still encounter the above issue without it.

**Comment by tatsuhiro-t** [comment]
I manged fix the remaining issue with the following patch.  The patch must be applied after #3091 applied.

```patch
From 517e03e2741931f97023a6a228957e0572efa5db Mon Sep 17 00:00:00 2001
From: Tatsuhiro Tsujikawa <tatsuhiro.t@gmail.com>
Date: Sat, 1 Apr 2017 11:54:39 +0900
Subject: [PATCH] Restore s->early_data_state with the original value

---
 ssl/ssl_lib.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/ssl/ssl_lib.c b/ssl/ssl_lib.c
index a76ee4068..f707b8b8e 100644
--- a/ssl/ssl_lib.c
+++ b/ssl/ssl_lib.c
@@ -1830,12 +1830,14 @@ int SSL_write_early_data(SSL *s, const void *buf, size_t num, size_t *written)
         return ret;
 
     case SSL_EARLY_DATA_FINISHED_READING:
-    case SSL_EARLY_DATA_READ_RETRY:
+    case SSL_EARLY_DATA_READ_RETRY: {
+        int early_data_state = s->early_data_state;
         /* We are a server writing to an unauthenticated client */
         s->early_data_state = SSL_EARLY_DATA_UNAUTH_WRITING;
         ret = SSL_write_ex(s, buf, num, written);
-        s->early_data_state = SSL_EARLY_DATA_READ_RETRY;
+        s->early_data_state = early_data_state;
         return ret;
+    }
 
     default:
         SSLerr(SSL_F_SSL_WRITE_EARLY_DATA, ERR_R_SHOULD_NOT_HAVE_BEEN_CALLED);
-- 
2.11.0

```

The fix is simple; just restore the original early_data_state value.  Feel free to integrate this to #3091.

**Comment by mattcaswell** [comment]
@tatsuhiro-t great thanks! Would you mind signing our CLA though so I can incorporate that patch?

https://www.openssl.org/policies/cla.html

**Comment by tatsuhiro-t** [comment]
Thanks.  I'll sign the CLA.

**Comment by tatsuhiro-t** [comment]
Alright, I signed CLA, and e-mail dispatched.

**Comment by mattcaswell** [comment]
Thanks. I can confirm we have received and recorded it.

**Comment by mattcaswell** [comment]
This should all be sorted now. I just pushed the remaining fixes to master.
